### PR TITLE
remove dependency of urlencoding crate

### PR DESCRIPTION
**Pull request by sreeo**
Fixes #
#1590 
Design discussion issue (if applicable) #

## Changes
Remove urlencoding dependency.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by cijothomas** [comment]
Fixes https://github.com/open-telemetry/opentelemetry-rust/issues/1590

File: opentelemetry/src/baggage.rs [hunk 1]
```diff
@@ -282,10 +282,29 @@ impl FromIterator<KeyValueMetadata> for Baggage {
     }
 }
 
+fn encode(s: Cow<'_, str>) -> String {
```
Comment (lalitb) [inline opentelemetry/src/baggage.rs:285]: nit - can this function take `&str` as it is not modifying the input string?

File: opentelemetry/src/baggage.rs [hunk 2]
```diff
@@ -282,10 +282,29 @@ impl FromIterator<KeyValueMetadata> for Baggage {
...
+    let special_characters = ['.', '-', '_', '~'];
+    let mut encoded_string = String::new();
```
Comment (lalitb) [inline opentelemetry/src/baggage.rs:287]: nit - we can preallocate the encoded string with the required length `s.len()`?

File: opentelemetry/src/baggage.rs [hunk 3]
```diff
@@ -282,10 +282,29 @@ impl FromIterator<KeyValueMetadata> for Baggage {
...
+
+    for char in s.as_ref().chars() {
+        if char.is_ascii_alphanumeric() || special_characters.contains(&char) {
+            encoded_string.push(char);
+        } else if char == ' ' {
+            encoded_string.push_str("%20");
+        } else {
+            // Convert to a two-digit hexadecimal representation
+            for byte in char.to_string().as_bytes() {
```
Comment (lalitb) [inline opentelemetry/src/baggage.rs:296]: That seems inefficient, converting char to string, and then to bytes. Instead, can we directly iterate over the original string as bytes?

File: opentelemetry/src/baggage.rs (hunk 1 above)
Comment (sreeo) [reply opentelemetry/src/baggage.rs]: yep

File: opentelemetry/src/baggage.rs (hunk 2 above)
Comment (sreeo) [reply opentelemetry/src/baggage.rs]: yep

File: opentelemetry/src/baggage.rs (hunk 3 above)
Comment (sreeo) [reply opentelemetry/src/baggage.rs]: yeah converting to char makes it easy to encode non ascii chars which could be multi byte.

Comment (lalitb) [reply opentelemetry/src/baggage.rs]: I meant something like this (not tested):
```rust
  for byte in s.as_bytes() {
        match *byte {
                    //alphanumeric...
                   b'a'.. b'z' | b 'A'.. b'Z' | ... => encoded_string.push(*byte as char);
                   b' ' =>  encoded_string.push_str("%20"),
                    //percentile encode all other type
                   _  =>  encoded_string.push_str(&format!("%{:02X}", byte)),
        }
  }
```

**Review by hdost — APPROVED** [review]

File: opentelemetry/src/baggage.rs (hunk 3 above)
Comment (sreeo) [reply opentelemetry/src/baggage.rs]: hmm, let me try this out.

File: opentelemetry/src/baggage.rs [hunk 4]
```diff
@@ -282,10 +281,53 @@ impl FromIterator<KeyValueMetadata> for Baggage {
...
+                    i = end - 1;
+                }
+            }
+        }
+
+        i += 1;
+    }
+
+    encoded_string
+}
```
Comment (lalitb) [inline opentelemetry/src/baggage.rs:325]: @sreeo Not sure if I am missing something, but we don't need separate handling for ASCII (non-alphanumeric) and UTF-8. Both need to be byte-wise percentile-encoded. I tested your unit-test examples on something I suggested earlier, and seems to work - https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1f433e6f80889b0680bab004f926ecaa.  Though it is difficult to estimate the encoded size, it should be fine to have initially set to the original string size, and not over-estimate. Sorry for being bit critical from perf prospective here :)

Comment (sreeo) [reply opentelemetry/src/baggage.rs:297]: No worries, your suggested code is simpler than the current one and passes all tests. Replaced it.

**Review by lalitb — APPROVED** [review]
Thanks for the fix.

**Pull request merged**